aten/src/ATen/core/PythonFallbackKernel.cpp

```diff
@@ -82,14 +88,33 @@ void pythonFallback(const c10::OperatorHandle& op, torch::jit::Stack* stack) {
         if (nv.isNone()) {
           continue;
         }
-        auto* interpreter = nv.unsafeToTensorImpl()->pyobj_slot()->pyobj_interpreter();
-        if (interpreter) {
-          (*interpreter)->dispatch(op, stack);
-          return;
+
+        auto* t = nv.unsafeToTensorImpl();
+        if (t->key_set().has(c10::DispatchKey::Python)) {
+          tensors_with_python_key_present = true;
+        }
+
+        auto* t_interpreter = t->pyobj_slot()->pyobj_interpreter();
+        if (!interpreter && t_interpreter) {
+          interpreter = t_interpreter;
         }
       }
     }
   }
+
+  if (interpreter) {
+    if (tensors_with_python_key_present) {
+      (*interpreter)->dispatch(op, stack);
+    } else {
+      // At this point, there are no modes in the stack and no tensors with the python key.
+      // so remove and disable the python key before redispatching.
+      c10::impl::ExcludeDispatchKeyGuard no_python_guard(c10::DispatchKey::Python);
```

Comment:
I guess alternatively, instead of adding it from the exclude set, we could remove it from the include set, and that would make it possible for someone in theory to push another mode somewhere else (though that seems unlikely?)

Reply:
I think the exclude guard here is actually... unnecessary? (does the test fail if you remove it?)

(1) `redispatchBoxed() will ignore the current include/exclude TLS, and just directly dispatch to the next key in the keyset that you passed in (which is almost guaranteed here to be one of the backend keys, CPU/CUDA/BackendSelect)

(2) The TLS will only matter in the case where the next function the dispatcher hits is composite, and invokes other ATen ops. Although... if we hit another op and end up going back to the python key, we should end up just re-triggering this logic.

Reply:
If I remove the exclude guard, I run into:
```
RuntimeError: 0 INTERNAL ASSERT FAILED at "/home/jbschlosser/branches/misc_fixes/aten/src/ATen/core/PythonFallbackKernel.cpp":123, please report a bug to PyTorch. Hit Python dispatch key but no arguments had PyInterpreter (no tensor args?)
```

and I don't quite understand why. are you sure the TLS is ignored by `redispatchBoxed()`?

Reply:
okay, clone() is composite explicit, so without the exclude guard, we'll redispatch into the python fallback for `empty_strided()` (which has no tensor inputs and thus no interpreters).

Reply:
Updated to remove from the include set instead!